### DISCUSSION TITLE
Can Dosemu2 run on a Raspberry Pi?

### DISCUSSION THREAD

**Discussion by baruchel** [dosemu2/dosemu2#1788] [Q&A]
Trying to compile Dosemu2 on a Rpi400, I couldn't manage to compile Fdpp. Did someone manage to compile the whole stuff (fdpp+dosemu2) on Raspbian?

**Comment by c-logic** [comment]
My latest status is :
Not without an additional emulation layer for x86 guests on the Cortex-A.

**Reply by stsp** [reply]
Something like qemu-user?
Does this even work?

**Comment by stsp** [comment]
I think building fdpp for arm
would be relatively simple.
Same byte-ordering as x86,
allows unaligned accesses
these days. The asm part
of fdpp should just be
cross-compiled with nasm
to x86 code.

I actually have plenty of ARM
boards around, so I could as
well try that myself.
But fdpp is only a small part
of the puzzle. I don't expect
building dosemu2 for arm would
be simple...
But its not impossible, as we
already have the CPU interpreter
for both RM and PM code.
So I don't anticipate any major
show-stopper here, besides the
lack of an user base for that.

**Comment by andrewbird** [comment]
I once wrote a patch to convert the dosemu driver files to nasm from gnu assembler, but the loss of the ability to include the C headers directly to share the #defines was a major drawback.

**Reply by stsp** [reply]
Which dosemu driver files are
still in nasm?
Why include is a problem if we
run cpp beforehand?

**Reply by andrewbird** [reply]
There are no dosemu driver files that use nasm, converting to it was my goal.

**Reply by stsp** [reply]
But if you mean that you
converted them, then they
still use includes, so I am
not sure what king of drawback
you mention.
Also I can't see in the log that
it was you, but perhaps it was
long ago. :)

**Reply by andrewbird** [reply]
Sorry I should have said, it was a patch that I never submitted because of the awkwardness with the includes.

**Reply by stsp** [reply]
This is how I interpreted that initially.
But makes no sense to me as I don't
see the nasm-coded drivers (and
haven't seen such for as long as can
remember) and don't see the include
problem you refer to.
If I try really hard, I can vaguely recall
that Bart made such conversion
somewhere in dosemu-0.x days.
Which is over 20 years ago.

**Reply by andrewbird** [reply]
The patch was to convert the gnu assembler to nasm, not the other way around. I hoped it would ease compilation on non x86 linux and perhaps make use of nasm's features including structure support.

**Reply by stsp** [reply]
Arrgh, okey. :)
This is indeed something interesting
to discuss, but as it was done the
other way around before, this should
be thoroughly considered.

**Reply by andrewbird** [reply]
I looked but I couldn't find any conversion from nasm, but there was one from [as86](https://wiki.osdev.org/AS86) by @bartoldeman to gas. Is this the commit you refer to 1361b24dd745d3862945dffb61425030be056885?

**Reply by stsp** [reply]
Likely so.
But it seems quite close
to a nasm syntax, isn't it?

**Reply by andrewbird** [reply]
Very similar apart from the comment character and segment override method. I also note that previously they were running the C preprocessor on nasm's input file as you mentioned before. Looking at the change log entry for Bart's conversion, I can only surmise that the conversion was done purely to move to something more modern, has the C preprocessor functionality and was available everywhere.

**Reply by stsp** [reply]
Yes, I am pretty sure the reason
was to drop the as86 dep as it was
probably fading from distros.
nasm is not fading and is also
already needed for fdpp anyway,
so maybe its not a big deal.
But I still don't know a compelling
reason.

**Reply by andrewbird** [reply]
> But I still don't know a compelling
reason.

Me neither and I still don't see a good solution to sharing the headers. At one time this tool https://github.com/JonathonReinhart/h2incn  looked to be going to make it into nasm's contrib, but development stalled https://forum.nasm.us/index.php?topic=1149.0.

The best I could up with was to process the headers that we need into nasm includes, but seems a little fragile. I still remember all the troubles with Make parallelism regarding bios.

#### src/include/Makefile
```make
%.inc: %.h                                                                      
        sed 's/^#define/%define/' $< | gcc -I. -D__ASSEMBLER__ -E -x c - | sed 's/^#.*//' > $@
```
Perhaps if interest for compilation on Raspberry Pi becomes a thing we can revisit, but for now let's not bother.

**Reply by stsp** [reply]
I don't think its even related.
When compiling for arm, we'll
still be using gas-i386.

**Reply by andrewbird** [reply]
Oh really I imagined that under arm the i386 tools wouldn't be available except if somebody bothered to create a cross compiler. So all good then,  sorry for the noise!

**Reply by stsp** [reply]
nasm should likely also be regarded
as a cross-compiler there. :)
I haven't checked, but I really suppose
gnu cross-tools are widely available
around distros, whereas cross nasm -
unlikely.

**Reply by ecm-pushbx** [reply]
NASM is very portable, that's one of its core qualities.

**Reply by stsp** [reply]
But have you checked that its
packaged in ARM releases of
fedora, ubuntu, debian etc?

**Reply by andrewbird** [reply]
To me it looks like Raspbian does not have the gnu assembler for i386, but equally it doesn't seem to have a package at all for nasm. Perhaps that's why @baruchel couldn't build FDPP.


I looked here http://archive.raspbian.org/multiarch/dists/buster/main/binary-arm64

Package: binutils-multiarch
Source: binutils
Version: 2.31.1-16+rpi2
Architecture: arm64
Maintainer: Matthias Klose <doko@debian.org>
Installed-Size: 9887
Depends: libc6 (>= 2.17), libfl2 (>= 2.5.33), zlib1g (>= 1:1.2.0), binutils (= 2.31.1-16+rpi2), binutils-common (= 2.31.1-16+rpi2)
Homepage: https://www.gnu.org/software/binutils/
Priority: optional
Section: devel
Filename: pool/main/b/binutils/binutils-multiarch_2.31.1-16+rpi2_arm64.deb
Size: 1642892
SHA256: 7d8328fc4964958db798a5d1042bbe3578e448c098f24988538525aa321577ba
SHA1: e0e55d63ba484674d70b79ab468ae37becc29aaa
MD5sum: ff67136a325179ab34041da1ca41ec3e
Description: Binary utilities that support multi-arch targets
 The programs in this package are used to manipulate binary and object
 files that may have been created on other architectures.  This package
 is primarily for multi-architecture developers and cross-compilers and
 is not needed by normal users or developers.  **Note that a cross-assembling
 version of gas is not included in this package, just the binary utilities.**
 .
 NORMAL USERS SHOULD NOT INSTALL THIS PACKAGE.  It's meant only for those
 requiring support for reading info from binaries from other architectures.

**Reply by stsp** [reply]
Maybe a bit weird, but I would
expect fdpp to be more easily
cross-compiled on x86_64 to
arm, than the other way around.
It has almost no deps so if you
install a C++-enabled cross
toolchain on x86_64, maybe
you'll be able to build.
But likely its quite useless as
you can't do the same with dosemu.

**Reply by stsp** [reply]
fdpp now builds on arm.
Wasn't difficult at all because
llvm tools (lld in this case) are
multi-target. So the same tool
can be used for native and
cross compilation.
Perhaps in dosemu2 we should
abandon the silly binutils finally,
and get cross-compilation for free.
Same applies to gcc vs clang.
Looks like a killer feature actually.
I wonder if GNU ppl are even
willing to compete or given up.

**Reply by stsp** [reply]
That trick fails on focal...
Andrew, could you please check
on Jammy? Does fdpp build?
I only have kinetic here and no
lp credentials. :)

**Reply by stsp** [reply]
OK LP already built everything
on jammy and kinetic, so only
focal has buggy lld.
I wonder if we should drop
focal support or do something
else...

**Reply by andrewbird** [reply]
> I wonder if we should drop
focal support 

I imagine many people could be upset as it's only two years old.

**Reply by stsp** [reply]
Yes but I believe this should
be handled via a bug report
for llvm-10. Since fdpp doesn't
even use any build system
(other than gnu make), I don't
want to dive into a cross-compilation
hell. Especially since in fdpp we
use clang anyway, so there is no
point to not use this cool feature
of the llvm stack.

**Reply by stsp** [reply]
> I imagine many people could be upset as it's only two years old.

I added the work-around to use
ld when lld is too old. But I doubt
its going to help because it will
still fail on arm.
Does anyone know if in debian
recipe it is possible to specify
different platforms for different
ubuntu versions? I.e. don't build
arm package on focal.
@skitt do you know if that is possible?

**Reply by stsp** [reply]
Also it seems focal does have an
lld-13 package in updates.
If we could require lld-13 for focal
and lld for the rest, that would also
allow to implement a work-around.

**Comment by stuaxo** [comment]
I tried this on a Pi I had to hand, that has debian bullseye, aarch64 and nasm seems to install there at least.

```bash
$ lsb_release -a
No LSB modules are available.
Distributor ID:	Debian
Description:	Debian GNU/Linux 11 (bullseye)
Release:	11
Codename:	bullseye
```

```
$ sudo apt install nasm
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
The following NEW packages will be installed:
  nasm
0 upgraded, 1 newly installed, 0 to remove and 30 not upgraded.
Need to get 408 kB of archives.
After this operation, 3,345 kB of additional disk space will be used.
Get:1 http://deb.debian.org/debian bullseye/main arm64 nasm arm64 2.15.05-1 [408 kB]
Fetched 408 kB in 0s (2,309 kB/s)
Selecting previously unselected package nasm.
(Reading database ... 114209 files and directories currently installed.)
Preparing to unpack .../nasm_2.15.05-1_arm64.deb ...
Unpacking nasm (2.15.05-1) ...
Setting up nasm (2.15.05-1) ...
Processing triggers for man-db (2.9.4-2) ...
Scanning processes...                                                                                                                                                                                                
Scanning processor microcode...                                                                                                                                                                                      
Scanning linux images...                                                                                                                                                                                             

Running kernel seems to be up-to-date.

Failed to check for processor microcode upgrades.

No services need to be restarted.

No containers need to be restarted.

No user sessions are running outdated binaries.
```

**Reply by andrewbird** [reply]
Cool thanks, at least Debian proper has it then.

**Comment by stsp** [comment]
So.
fdpp now builds for arm
(should even on LP I think).
dosemu2 may require some
work, namely I see it using
the asm-based bitops and
also sigcontext is incompatible.
But I don't expect too many
problems. I believe dosemu2
is in a good shape these days
and can be ported with sufficiently
few efforts.

**Reply by andrewbird** [reply]
I updated my desktop machine to jammy now, fdpp builds fine, but I'm seeing this problem with dosemu2. Is there some other package I need as I didn't see any recent changes to debian/control?
~~~
make[2]: Entering directory '/clients/common/dosemu2.git/src/base/video'
make[2]: *** No rule to make target '/usr/lib/llvm-10/lib/clang/10.0.0/include/stddef.h', needed by 'text.o'. Stop.
make[2]: Leaving directory '/clients/common/dosemu2.git/src/base/video'
make[1]: *** [/clients/common/dosemu2.git/src/arch/linux/Makefile.main:104: base/video] Error 2
make[1]: Leaving directory '/clients/common/dosemu2.git/src'
~~~

**Reply by stsp** [reply]
make clean

**Reply by andrewbird** [reply]
I was already using `make distclean` which didn't fix the problem, so trying again now with `make clean` to see if there's any difference.

**Reply by andrewbird** [reply]
Thanks, `make clean` worked fine.

**Comment by stsp** [comment]
@baruchel ARM builds are now in
PPA. They are very experimental for
now, so you are welcome to experiment. :)
And let us know the results.

**Comment by stsp** [comment]
I also remember someone else
was recently trying to create the
buildroot environment with dosemu2
on arm. But I can't find that thread.
Does anyone remember that?

**Reply by stuaxo** [reply]
I have a really vague memory of playing with buildroot - and may well have been using buildroot and an ARM vm in Qemu).

I don't have the same computer has about 1 1/2 years ago, so it's a bit tricky to check,

**Reply by stsp** [reply]
No, it wasn't you, but thanks for
a reminder. This time I've found that
arm thread:
https://github.com/dosemu2/dosemu2/issues/732
Maybe the guy from there is still with
us, as it was the most recent thread
of all.
